### PR TITLE
[Merged by Bors] - EnvironmentMapLight, BRDF Improvements

### PR DESCRIPTION
**Pull request by JMS55**
(Before)
image
(After)
image
image

# Objective
- Improve lighting; especially reflections.
- Closes https://github.com/bevyengine/bevy/issues/4581.

## Solution
- Implement environment maps, providing better ambient light.
- Add microfacet multibounce approximation for specular highlights from Filament.
- Occlusion is no longer incorrectly applied to direct lighting. It now only applies to diffuse indirect light. Unsure if it's also supposed to apply to specular indirect light - the glTF specification just says "indirect light". In the case of ambient occlusion, for instance, that's usually only calculated as diffuse though. For now, I'm choosing to apply this just to indirect diffuse light, and not specular.
- Modified the PBR example to use an environment map, and have labels.
- Added `FallbackImageCubemap`.

## Implementation
- IBL technique references can be found in environment_map.wgsl.
- It's more accurate to use a LUT for the scale/bias. Filament has a good reference on generating this LUT. For now, I just used an analytic approximation.
 - For now, environment maps must first be prefiltered outside of bevy using a 3rd party tool. See the `EnvironmentMap` documentation.
- Eventually, we should have our own prefiltering code, so that we can have dynamically changing environment maps, as well as let users drop in an HDR image and use asset preprocessing to create the needed textures using only bevy. 

---

## Changelog
- Added an `EnvironmentMapLight` camera component that adds additional ambient light to a scene.
- StandardMaterials will now appear brighter and more saturated at high roughness, due to internal material changes. This is more physically correct.
- Fixed StandardMaterial occlusion being incorrectly applied to direct lighting.
- Added `FallbackImageCubemap`.

### CODE REVIEW

File: crates/bevy_pbr/src/environment_map/mod.rs [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+/// an environment cubemap texture.
+///
+/// The environment map must be prefiltered into a diffuse and specular map based on the
+/// [split-sum approximation](https://cdn2.unrealengine.com/Resources/files/2013SiggraphPresentationsNotes-26915738.pdf).
+/// The specular map must have exactly 11 mips [0, 10].
+///
+/// To prefilter your environment map, you can use KhronosGroup's [glTF-IBL-Sampler](https://github.com/KhronosGroup/glTF-IBL-Sampler).
+/// The diffuse map uses the Lambertian distribution, and the specular map uses the GGX distribution.
+///
+/// KhronoGroup also has several prefiltered environment maps that can be found [here](https://github.com/KhronosGroup/glTF-Sample-Environments).
```
Comment (IceSentry) [inline crates/bevy_pbr/src/environment_map/mod.rs:64]:
```suggestion
/// To prefilter your environment map, you can use `KhronosGroup`'s [glTF-IBL-Sampler](https://github.com/KhronosGroup/glTF-IBL-Sampler).
/// The diffuse map uses the Lambertian distribution, and the specular map uses the GGX distribution.
///
/// `KhronoGroup` also has several prefiltered environment maps that can be found [here](https://github.com/KhronosGroup/glTF-Sample-Environments).
```

This lint is a bit annoying, but this is why CI is failing

**Comment by IceSentry** [comment]
Can you link the scene you used in the second screenshot? I know it was one of the gltf samples, but just to make it easier for people to reproduce

File: examples/3d/pbr.rs [hunk 2]
```diff
@@ -69,14 +72,115 @@ fn setup(
...
+                ..default()
+            },
+            ..default()
+        }),
+        EnvironmentMapLabel,
+    ));
+
+    // camera
+    let environment_map = EnvironmentMap {
+        diffuse_map: asset_server.load("environment_maps/pisa_diffuse.ktx2"),
```
Comment (IceSentry) [inline examples/3d/pbr.rs:145]: Since this example requires the `ktx2` feature, I wonder if it's possible to tell cargo to use it automatically for this example

Comment (IceSentry) [reply examples/3d/pbr.rs:145]: To be clear, I know it's a required feature, but cargo still asked me to enter it manually

**Review by IceSentry — APPROVED** [review]
Overall, code looks good, I can't comment on the math in the shader though. I'll try to go over the reference material at some point.

File: examples/3d/pbr.rs (hunk 2 above)
Comment (JMS55) [reply examples/3d/pbr.rs:145]: Not possible sadly https://github.com/rust-lang/cargo/issues/4663

Comment (IceSentry) [reply examples/3d/pbr.rs:145]: 😢

**Comment by JMS55** [comment]
> Can you link the scene you used in the second screenshot? I know it was one of the gltf samples, but just to make it easier for people to reproduce

The scene is at the following link. If you want to reproduce the screenshot exactly, make sure to turn off all direct lighting, ambient lighting, and tonemapping, and enable hdr. https://github.com/KhronosGroup/glTF-Sample-Models/tree/master/2.0/MetalRoughSpheres.

> Overall, code looks good, I can't comment on the math in the shader though. I'll try to go over the reference material at some point.

Shader math is exactly taken from the multiscattering paper, and unreal's mobile rendering blog post. I've changed basically nothing from them. Should be entirely correct, assuming I didn't make any more copy paste errors :)

For the most part I'm happy enough with the Rust code, except for the DrawFunction thing. Having to make a second DrawFunction, and have an ugly if statement to decide between them is kinda gross. Doesn't seem like a great way to do it, and I'm open to ideas on how to improve it. This was sadly the best I could come up with.

**Comment by JMS55** [comment]
I'm probably going to rename EnvironmentMap to EnvironmentMapLight, to indicate it's a lighting component.

**Comment by JMS55** [comment]
Note that the second screenshot has changed since the above comments. It's now using tonemapping and direct lighting.

**Review by superdump — CHANGES_REQUESTED** [review]
For now, add this to the view bind group. I know it's cumbersome and not modular, but rebinding a bind group causes all bind groups after it to be rebound so I feel the env map belongs in the view bind group. We'll make the view bind group more modular later.

Also, the example assets are huge. The largest asset we have in the repo is just over 3MB, and everything else is 1MB or smaller. If there is no compressed format support, then perhaps at least we could use png or zstd ktx2, and reduce the resolution until the size is acceptable. A quality decrease is ok as examples are meant to be illustrative. For release articles we can use higher-quality assets.

File: crates/bevy_pbr/src/environment_map/mod.rs [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+/// The specular map must have exactly 11 mips [0, 10].
+///
+/// To prefilter your environment map, you can use `KhronosGroup`'s [glTF-IBL-Sampler](https://github.com/KhronosGroup/glTF-IBL-Sampler).
+/// The diffuse map uses the Lambertian distribution, and the specular map uses the GGX distribution.
+///
+/// `KhronoGroup` also has several prefiltered environment maps that can be found [here](https://github.com/KhronosGroup/glTF-Sample-Environments).
+#[derive(Component, Reflect, Clone)]
+pub struct EnvironmentMapLight {
+    pub diffuse_map: Handle<Image>,
+    pub specular_map: Handle<Image>,
```
[the remark this answers is not in the document]
Comment (JMS55) [reply crates/bevy_pbr/src/environment_map/mod.rs:66]: They correspond to different parts of the BRDF calculations. You need both - it's not like the PBR material with different parameters. These are two parts of the same thing.

File: crates/bevy_pbr/src/environment_map/environment_map.wgsl [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+
+    // Multiscattering approximation: https://www.jcgt.org/published/0008/01/03/paper.pdf
+    // Useful reference: https://bruop.github.io/ibl
+    let Fr = max(vec3(1.0 - roughness), F0) - F0;
+    let kS = F0 + Fr * pow(1.0 - NdotV, 5.0);
+    let FssEss = kS * f_ab.x + f_ab.y;
+    let Ess = f_ab.x + f_ab.y;
+    let Ems = 1.0 - Ess;
+    let Favg = F0 + (1.0 - F0) / 21.0;
+    let Fms = FssEss * Favg / (1.0 - Ems * Favg);
```
Comment (cart) [inline crates/bevy_pbr/src/environment_map/environment_map.wgsl:34]: Shouldn't this be `let Fms = FssEss * Favg / (1.0 - Ess * Favg);` (note the `Ess` instead of `Ems`)

Comment (cart) [reply crates/bevy_pbr/src/environment_map/environment_map.wgsl:34]: image

Comment (cart) [reply crates/bevy_pbr/src/environment_map/environment_map.wgsl:34]: Doing the swap definitely breaks some things. So either this was an intentional change to the algorithm to fix something (which I admittedly don't fully understand yet), or other parts of the feature are broken :)
image

Comment (JMS55) [reply crates/bevy_pbr/src/environment_map/environment_map.wgsl:34]: Note that the paper does `FssEss * Favg/(1 - (1 - Ess) * Favg)`.

I did `Fms = FssEss * Favg / (1.0 - Ems * Favg)`, because `Ems = 1.0 - Ess`.

Comment (cart) [reply crates/bevy_pbr/src/environment_map/environment_map.wgsl:34]: Ahhh yup that does explain it. Carry on :)

File: crates/bevy_pbr/src/render/pbr_lighting.wgsl [hunk 5]
```diff
@@ -101,13 +101,17 @@ fn fresnel(f0: vec3<f32>, LoH: f32) -> vec3<f32> {
...
+fn specular(f0: vec3<f32>, roughness: f32, h: vec3<f32>, NoV: f32, NoL: f32, NoH: f32, LoH: f32, specularIntensity: f32, f_ab: vec2<f32>) -> vec3<f32> {
     let D = D_GGX(roughness, NoH, h);
     let V = V_SmithGGXCorrelated(roughness, NoV, NoL);
     let F = fresnel(f0, LoH);
 
-    return (specularIntensity * D * V) * F;
+    var Fr = (specularIntensity * D * V) * F;
+
+    // Multiscattering approximation: https://google.github.io/filament/Filament.html#listing_energycompensationimpl
+    Fr *= 1.0 + f0 * (1.0 / f_ab.x - 1.0);
```
Comment (IceSentry) [inline crates/bevy_pbr/src/render/pbr_lighting.wgsl:112]: the link does:
```hlsl
vec3 energyCompensation = 1.0 + f0 * (1.0 / dfg.y - 1.0);
```

I'm not sure what dfg here is in this context, I assume it's the same as `f_ab`, but it uses `.y` and you used `.x`. Was this intentional?

Comment (JMS55) [reply crates/bevy_pbr/src/render/pbr_lighting.wgsl:112]: It looks very wrong when swapped to `.y`. I forget why it's backwards vs filament, maybe filament has theirs backwards relative to unreal and others. I'm pretty sure it's correct, though.

File: crates/bevy_pbr/src/environment_map/environment_map.wgsl [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+    let Edss = 1.0 - (FssEss + Fms * Ems);
+    let kD = diffuse_color * Edss;
```
Comment (IceSentry) [inline crates/bevy_pbr/src/environment_map/environment_map.wgsl:36]: I personally found it much easier to follow when keeping the comments from the reference

```suggestion
    // Split-sum approximation for image based lighting: https://cdn2.unrealengine.com/Resources/files/2013SiggraphPresentationsNotes-26915738.pdf
    let smallest_specular_mip_level = textureNumLevels(environment_map_specular) - 1i;
    let radiance_level = perceptual_roughness * f32(smallest_specular_mip_level);
    
    // Cosine-weighted irradiance
    let irradiance = textureSample(environment_map_diffuse, environment_map_sampler, N).rgb;
    // Prefiltered radiance
    let radiance = textureSampleLevel(environment_map_specular, environment_map_sampler, R, radiance_level).rgb;

    // Multiscattering approximation: https://www.jcgt.org/published/0008/01/03/paper.pdf
    // Useful reference: https://bruop.github.io/ibl
    
    // Roughness dependent fresnel
    let Fr = max(vec3(1.0 - roughness), F0) - F0;
    let kS = F0 + Fr * pow(1.0 - NdotV, 5.0);
    let FssEss = kS * f_ab.x + f_ab.y;
    
    // Multiple scattering
    let Ess = f_ab.x + f_ab.y;
    let Ems = 1.0 - Ess;
    let Favg = F0 + (1.0 - F0) / 21.0;
    let Fms = FssEss * Favg / (1.0 - Ems * Favg);
    
    // Dielectrics
    let Edss = 1.0 - (FssEss + Fms * Ems);
    let kD = diffuse_color * Edss;
```

File: crates/bevy_pbr/src/render/mesh.rs [hunk 7]
```diff
@@ -412,10 +412,16 @@ impl FromWorld for MeshPipeline {
                     count: None,
                 },
             ];
+
+            // EnvironmentMapLight
+            let environment_map_entries =
+                environment_map::get_bind_group_layout_entries([11, 12, 13]);
```
Comment (IceSentry) [inline crates/bevy_pbr/src/render/mesh.rs:418]: I really like this pattern, we should probably do this for other parts of the engine. The prepass should probably do that at least.

File: crates/bevy_pbr/src/render/mesh.rs [hunk 8]
```diff
@@ -741,10 +749,13 @@ impl SpecializedMeshPipeline for MeshPipeline {
...
-        } else {
-            TextureFormat::bevy_default()
+        if key.contains(MeshPipelineKey::ENVIRONMENT_MAP) {
+            shader_defs.push("ENVIRONMENT_MAP".into());
+        }
+
+        let format = match key.contains(MeshPipelineKey::HDR) {
+            true => ViewTarget::TEXTURE_FORMAT_HDR,
+            false => TextureFormat::bevy_default(),
         };
```
Comment (IceSentry) [inline crates/bevy_pbr/src/render/mesh.rs:759]: why was this changed?

**Review by IceSentry — APPROVED** [review]
LGTM

I doubled checked the math from the linked references and everything seemed correct.

Couple nits but nothing blocking otherwise.

File: crates/bevy_pbr/src/render/mesh.rs (hunk 8 above)
Comment (JMS55) [reply crates/bevy_pbr/src/render/mesh.rs:759]: I thought it looked better, it keeps the two texture formats right next to each other instead of separated like an if statement. I can revert it though.

Comment (IceSentry) [reply crates/bevy_pbr/src/render/mesh.rs:759]: While you aren't the only one that likes it, cart doesn't and has asked to not use it in the past. I thought we had a clippy rule for that. I don't really mind personally.

Comment (JMS55) [reply crates/bevy_pbr/src/render/mesh.rs]: Reverted.

File: examples/3d/pbr.rs [hunk 9]
```diff
@@ -69,14 +72,115 @@ fn setup(
...
+    if asset_server.get_load_state(&handles.handles.diffuse_map) == LoadState::Loaded
+        && asset_server.get_load_state(&handles.handles.specular_map) == LoadState::Loaded
+    {
+        if let Ok(label_entity) = label_query.get_single() {
+            commands.entity(label_entity).despawn();
+        }
+    }
+}
+
+#[derive(Resource)]
```
Comment (cart) [inline examples/3d/pbr.rs:180]: We can (and I think should) simplify this by removing the resource:

```rust

    // camera
    commands.spawn((
        Camera3dBundle {
            transform: Transform::from_xyz(0.0, 0.0, 8.0).looking_at(Vec3::default(), Vec3::Y),
            projection: OrthographicProjection {
                scale: 0.01,
                ..default()
            }
            .into(),
            ..default()
        },
        EnvironmentMapLight {
            diffuse_map: asset_server.load("environment_maps/pisa_diffuse_zstd.ktx2"),
            specular_map: asset_server.load("environment_maps/pisa_specular_zstd.ktx2"),
        },
    ));
}

fn environment_map_load_finish(
    mut commands: Commands,
    asset_server: Res<AssetServer>,
    environment_maps: Query<&EnvironmentMapLight>,
    label_query: Query<Entity, With<EnvironmentMapLabel>>,
) {
    let Ok(environment_map) = environment_maps.get_single() else {
        return
    };
    if asset_server.get_load_state(&environment_map.diffuse_map) == LoadState::Loaded
        && asset_server.get_load_state(&environment_map.specular_map) == LoadState::Loaded
    {
        if let Ok(label_entity) = label_query.get_single() {
            commands.entity(label_entity).despawn();
        }
    }
}

#[derive(Component)]
struct EnvironmentMapLabel;
```

File: crates/bevy_pbr/src/render/pbr_lighting.wgsl [hunk 10]
```diff
@@ -131,14 +135,19 @@ fn Fd_Burley(roughness: f32, NoV: f32, NoL: f32, LoH: f32) -> f32 {
     return lightScatter * viewScatter * (1.0 / PI);
 }
 
-// From https://www.unrealengine.com/en-US/blog/physically-based-shading-on-mobile
-fn EnvBRDFApprox(f0: vec3<f32>, perceptual_roughness: f32, NoV: f32) -> vec3<f32> {
+// Scale/bias approximation
+// https://www.unrealengine.com/en-US/blog/physically-based-shading-on-mobile
+// TODO: Use a LUT (more accurate)
```
Comment (cart) [inline crates/bevy_pbr/src/render/pbr_lighting.wgsl:140]: Presumably we might encounter issues like this until we add this?
https://github.com/google/filament/discussions/4989

**Review by cart — APPROVED** [review]
This looks good to me! Scenes are going to look so much better now :)

File: crates/bevy_pbr/src/render/pbr_lighting.wgsl (hunk 10 above)
Comment (JMS55) [reply crates/bevy_pbr/src/render/pbr_lighting.wgsl:140]: Not sure. Haven't seen anything in testing though.

File: crates/bevy_pbr/src/render/pbr_ambient.wgsl [hunk 11]
```diff
@@ -12,8 +12,8 @@ fn ambient_light(
...
     occlusion: f32,
 ) -> vec3<f32> {
-    let diffuse_ambient = EnvBRDFApprox(diffuse_color, 1.0, NdotV);
-    let specular_ambient = EnvBRDFApprox(specular_color, perceptual_roughness, NdotV);
+    let diffuse_ambient = EnvBRDFApprox(diffuse_color, F_AB(1.0, NdotV)) * occlusion;
+    let specular_ambient = EnvBRDFApprox(specular_color, F_AB(perceptual_roughness, NdotV));
 
-    return (diffuse_ambient + specular_ambient) * lights.ambient_color.rgb * occlusion;
+    return (diffuse_ambient + specular_ambient) * lights.ambient_color.rgb;
 }
```
[the remark this answers is not in the document]
Comment (JMS55) [reply crates/bevy_pbr/src/render/pbr_ambient.wgsl:19]: Recently had to reinstall my OS, and forgot to reenable that setting :P

File: crates/bevy_pbr/src/environment_map/environment_map.wgsl [hunk 12]
```diff
@@ -0,0 +1,42 @@
...
+
+    var out: EnvironmentMapLight;
+    out.diffuse = (Fms * Ems + kD) * irradiance;
```
[the remark this answers is not in the document]
Comment (JMS55) [reply crates/bevy_pbr/src/environment_map/environment_map.wgsl:39]: Nope, just missed that I could do that. I mainly referenced that blog post for understanding, the code was taken from the original paper.

**Comment by JMS55** [comment]
Not sure how to fix the CI needing ktx2/zstd features :P

**Comment by IceSentry** [comment]
@JMS55 for the CI look at what the skybox example does in the top level Cargo.toml. You should be able to specify required features for an example.

**Comment by JMS55** [comment]
> @JMS55 for the CI look at what the skybox example does in the top level Cargo.toml. You should be able to specify required features for an example.

I did that already, it doesn't seem like enough for the CI :/

**Comment by superdump** [comment]
I think this could fix CI:
```diff
diff --git a/.github/workflows/ci.yml b/.github/workflows/ci.yml
index fce2860e5..f40993c7c 100644
--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -190,14 +190,14 @@ jobs:
       - name: Build bevy
         # this uses the same command as when running the example to ensure build is reused
         run: |
-          TRACE_CHROME=trace-alien_cake_addict.json CI_TESTING_CONFIG=.github/example-run/alien_cake_addict.ron cargo build --example alien_cake_addict --features "bevy_ci_testing,trace,trace_chrome"
+          TRACE_CHROME=trace-alien_cake_addict.json CI_TESTING_CONFIG=.github/example-run/alien_cake_addict.ron cargo build --example alien_cake_addict --features "bevy_ci_testing,trace,trace_chrome,ktx2,zstd"
       - name: Run examples
         run: |
           for example in .github/example-run/*.ron; do
             example_name=`basename $example .ron`
             echo -n $example_name > last_example_run
             echo "running $example_name - "`date`
-            time TRACE_CHROME=trace-$example_name.json CI_TESTING_CONFIG=$example xvfb-run cargo run --example $example_name --features "bevy_ci_testing,trace,trace_chrome"
+            time TRACE_CHROME=trace-$example_name.json CI_TESTING_CONFIG=$example xvfb-run cargo run --example $example_name --features "bevy_ci_testing,trace,trace_chrome,ktx2,zstd"
             sleep 10
           done
           zip traces.zip trace*.json
diff --git a/.github/workflows/validation-jobs.yml b/.github/workflows/validation-jobs.yml
index 4d306320d..3b36ab363 100644
--- a/.github/workflows/validation-jobs.yml
+++ b/.github/workflows/validation-jobs.yml
@@ -80,7 +80,7 @@ jobs:
         shell: bash
         # this uses the same command as when running the example to ensure build is reused
         run: |
-          WGPU_BACKEND=dx12 CI_TESTING_CONFIG=.github/example-run/alien_cake_addict.ron cargo build --example alien_cake_addict --features "bevy_ci_testing"
+          WGPU_BACKEND=dx12 CI_TESTING_CONFIG=.github/example-run/alien_cake_addict.ron cargo build --example alien_cake_addict --features "bevy_ci_testing,ktx2,zstd"
 
       - name: Run examples
         shell: bash
@@ -88,7 +88,7 @@ jobs:
           for example in .github/example-run/*.ron; do
             example_name=`basename $example .ron`
             echo "running $example_name - "`date`
-            time WGPU_BACKEND=dx12 CI_TESTING_CONFIG=$example cargo run --example $example_name --features "bevy_ci_testing"
+            time WGPU_BACKEND=dx12 CI_TESTING_CONFIG=$example cargo run --example $example_name --features "bevy_ci_testing,ktx2,zstd"
             sleep 10
           done
 
```
